vdc/scripts/users/user_data.py

```diff
@@ -0,0 +1,155 @@
+import shortuuid
+
+import google
+from google.cloud import storage
+from googleapiclient.errors import HttpError
+from globals import k8s, kube_client, gcloud_service
+from table import Table
+
+shortuuid.set_alphabet("0123456789abcdefghijkmnopqrstuvwxyz")
+
+
+def create_service_id():
+    return f'user-{shortuuid.uuid()}'
+
+
+def create_google_service_account(sa_name, google_project):
+    return gcloud_service.projects().serviceAccounts().create(
+            name=f'projects/{google_project}',
+            body={
+                "accountId": sa_name, "serviceAccount": {
+                    "displayName": "user"
+                }
+            }).execute()
+
+
+def delete_google_service_account(gsa_email, google_project):
+    return gcloud_service.projects().serviceAccounts().delete(name=f'projects/-/serviceAccounts/{gsa_email}').execute()
+
+
+def create_kube_service_acccount(namespace):
+    return k8s.create_namespaced_service_account(
+        namespace=namespace,
+        body=kube_client.V1ServiceAccount(
+            api_version='v1',
+            metadata=kube_client.V1ObjectMeta(
+                generate_name='user-',
+                annotations={
+                    "type": "user"
+                }
+            )
+        )
+    )
+
+
+def delete_kube_service_acccount(ksa_name, namespace):
+    return k8s.delete_namespaced_service_account(name=ksa_name, namespace=namespace, body={})
+
+
+def create_bucket(sa_name, gsa_email):
+    bucket = storage.Client().bucket(sa_name)
+    bucket.labels = {
+        'type': 'user',
+    }
+
+    bucket.create()
+
+    acl = bucket.acl
+    acl.user(gsa_email).grant_owner()
+
+    return bucket
+
+
+def delete_bucket(bucket_name):
+    return storage.Client().get_bucket(bucket_name).delete()
+
+
+def create_all(google_project, kube_namespace):
+    out = {}
+
+    ksa_response = create_kube_service_acccount(kube_namespace)
+    out['ksa_name'] = ksa_response.metadata.name
+
+    sa_name = create_service_id()
+
+    gs_response = create_google_service_account(sa_name, google_project)
+    out['gsa_email'] = gs_response['email']
+
+    create_bucket(sa_name, out['gsa_email'])
+    out['bucket_name'] = sa_name
+
+    return out
+
+
+def delete_all(user_obj, google_project='hail-vdc', kube_namespace='default'):
+    modified = 0
+    try:
+        delete_bucket(user_obj['bucket_name'])
+        modified += 1
+    except google.api_core.exceptions.NotFound:
+        pass
+
+    try:
+        delete_google_service_account(user_obj['gsa_email'], google_project)
+        modified += 1
+    except HttpError as e:
+        if e.resp.status != 404:
+            raise e
+
+    try:
+        delete_kube_service_acccount(user_obj['ksa_name'], kube_namespace)
+        modified += 1
+    except kube_client.rest.ApiException as e:
+        if e.status != 404:
+            raise e
+
+    if modified == 0:
+        return 404
```

Comment:
If modified !=3, I don't return 404 to indicate that at least one of the resources were found; I give this level of granularity to provide a message indicating that at least one resource was deleted, since partial resources may exist due to communication interruptions. I have tests written for this, such as https://github.com/hail-is/hail/pull/5618/files/2fb881589891defaac478d3b0fc1866dd370f781#diff-62f83960a423fd4a9ee5dbc963cb0219R81

If 0 resources were deleted (404), I indicate that [Nothing to changed ...](https://github.com/hail-is/hail/blob/2fb881589891defaac478d3b0fc1866dd370f781/vdc/scripts/users/user_data.py#L151), otherwise that we deleted the requested resources.

Reply:
ok.